{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start your mySQL Server from a terminal \n",
    "(if it isn't already running)\n",
    "\n",
    "<code>sudo docker start course-mysql</code>\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Don't forget that, if you use sqlMagic, you need to execute the connection lines in your Notebook!\n",
    "\n",
    "<pre>\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "</pre>\n",
    "\n",
    "## Create a new Python3 Jupyter Notebook in your Exam Answers folder\n",
    "\n",
    "Commit and push this Notebook to GitHub when you are finished.\n",
    "\n",
    "You must submit your answers to GitHub by 1800h Sept 14."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bash:\n",
    "\n",
    "# docker start course-mysql\n",
    "# mysql --local_infile=true -h localhost -P 3306 --protocol=TCP -u root -p "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "++\n",
       "||\n",
       "++\n",
       "++"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from sqlalchemy import URL, create_engine, MetaData\n",
    "from getpass import getpass\n",
    "\n",
    "url_object = URL.create(\n",
    "    \"mysql+pymysql\",\n",
    "    username=\"root\",\n",
    "    password= getpass(),  # plain (unescaped) text\n",
    "    host=\"localhost\", # Averiguar en la línea de comandos con \"status;\". Ni idea de porqué no hace ni puto caso al comando de antes.\n",
    "    database=\"mysql\"\n",
    ")\n",
    "\n",
    "engine = create_engine(url_object)\n",
    "\n",
    "metadata = MetaData() # Crea una variable para almacenar los metadatos de las tablas.\n",
    "\n",
    "# Inicia el interprete de pymysql.\n",
    "%load_ext sql \n",
    "\n",
    "# Realiza la conexión.\n",
    "%sql engine \n",
    "\n",
    "# Permite cargar archivos locales.\n",
    "%sql SET GLOBAL local_infile=true;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Data Files\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "def get_first_column(\n",
    "    file_path: str, \n",
    "    header: bool = False, \n",
    "    delimiter: str = '\\t', \n",
    "    quotechar: str = '\"'\n",
    "    ) -> list: \n",
    "    \"\"\"Return the first column of a data file in a single list. \n",
    "    Each value is imported as a string.\n",
    "\n",
    "    :param: file_name: str - Path to the file containing the data.\n",
    "    :param: header: bool - Default false. True if the file contains a \n",
    "    header and should be ignored. \n",
    "    :param: delimiter: str - Default tab (.tsv file). Delimiter for \n",
    "    the fields it the data file.\n",
    "    :param: quotechar: str - Default '\"'. Quoting for the imported data.\n",
    "    :return: list - List containing all the values from the first \n",
    "    column of the file.\n",
    "    \"\"\"    \n",
    "\n",
    "    import csv\n",
    "\n",
    "    first_column = []\n",
    "    rest = []\n",
    "\n",
    "    with open(file_path) as file:\n",
    "\n",
    "        if header: next(file) # Es un iterable así que nos saltamos la cabecera.\n",
    "        \n",
    "        file_reader = csv.reader(file, delimiter = delimiter, quotechar = quotechar)\n",
    "        \n",
    "        for first_column_values, *rest in file_reader: \n",
    "            # Itera por filas y automáticamente guarda el valor de cada \n",
    "            # campo en cada una de las variables.\n",
    "            first_column.append(first_column_values)\n",
    "            rest.clear() # Rest no lo itero así que solo guardará el último valor.\n",
    "\n",
    "    return first_column\n",
    "\n",
    "\n",
    "locus_Germplasm = get_first_column(\"Germplasm.tsv\", True)\n",
    "locus_LocusGene = get_first_column(\"LocusGene.tsv\", True)\n",
    "\n",
    "print(locus_Germplasm == locus_LocusGene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">2 rows affected.</span>"
      ],
      "text/plain": [
       "2 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">1 rows affected.</span>"
      ],
      "text/plain": [
       "1 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">5 rows affected.</span>"
      ],
      "text/plain": [
       "5 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>Tables_in_germplasm</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+---------------------+\n",
       "| Tables_in_germplasm |\n",
       "+---------------------+\n",
       "+---------------------+"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database germplasm;\n",
    "%sql create database germplasm;\n",
    "%sql show databases;\n",
    "%sql use germplasm;\n",
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password=getpass(),\n",
    "                             db='germplasm', \n",
    "                             charset='utf8mb4',\n",
    "                             local_infile=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        sql = \"\"\"create table germplasm(\n",
    "            locus VARCHAR(20) NOT NULL PRIMARY KEY,\n",
    "            germplasm VARCHAR(100) NOT NULL,\n",
    "            phenotype VARCHAR(1000) NOT NULL,\n",
    "            pubmed INTEGER NOT NULL);\"\"\"\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        sql = \"\"\"create table locus_gene(\n",
    "            locus VARCHAR(20) NOT NULL PRIMARY KEY,\n",
    "            gene VARCHAR(20) NOT NULL,\n",
    "            protein_lenght INTEGER NOT NULL)\"\"\"\n",
    "        cursor.execute(sql)\n",
    "\n",
    "    connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\\nDone\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">2 rows affected.</span>"
      ],
      "text/plain": [
       "2 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>Tables_in_germplasm</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>germplasm</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>locus_gene</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+---------------------+\n",
       "| Tables_in_germplasm |\n",
       "+---------------------+\n",
       "|      germplasm      |\n",
       "|      locus_gene     |\n",
       "+---------------------+"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show tables;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>germplasm</th>\n",
       "            <th>phenotype</th>\n",
       "            <th>pubmed</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+-------+-----------+-----------+--------+\n",
       "| locus | germplasm | phenotype | pubmed |\n",
       "+-------+-----------+-----------+--------+\n",
       "+-------+-----------+-----------+--------+"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>gene</th>\n",
       "            <th>protein_lenght</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+-------+------+----------------+\n",
       "| locus | gene | protein_lenght |\n",
       "+-------+------+----------------+\n",
       "+-------+------+----------------+"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password=getpass(),\n",
    "                             db='germplasm', \n",
    "                             charset='utf8mb4',\n",
    "                             local_infile=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        germplasm_path = \"/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv\"\n",
    "        locus_gene_path = \"/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/LocusGene.tsv\"\n",
    "\n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE germplasm\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = germplasm_path)\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE locus_gene\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = locus_gene_path)\n",
    "        cursor.execute(sql)\n",
    "\n",
    "        connection.commit()\n",
    "\n",
    "finally:\n",
    "    print(\"\\nDone\")\n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #  PARA EL PC PERSONAL\n",
    "\n",
    "# import pymysql.cursors\n",
    "\n",
    "# connection = pymysql.connect(host='localhost',\n",
    "#                              user='root',\n",
    "#                              password=getpass(),\n",
    "#                              db='germplasm', \n",
    "#                              charset='utf8mb4',\n",
    "#                              local_infile=True,\n",
    "#                              cursorclass=pymysql.cursors.DictCursor)\n",
    "# connection.autocommit = False\n",
    "\n",
    "# try:\n",
    "#     with connection.cursor() as cursor:\n",
    "\n",
    "#         germplasm_path = \"/home/jm/Course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv\"\n",
    "#         locus_gene_path = \"/home/jm/Course/Accelerated_Intro_WilkinsonExams/LocusGene.tsv\"\n",
    "\n",
    "#         sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "#         INTO TABLE germplasm\n",
    "#         FIELDS TERMINATED BY '\\t'\n",
    "#         LINES TERMINATED BY '\\n'\n",
    "#         IGNORE 1 LINES;\"\"\".format(path = germplasm_path)\n",
    "#         cursor.execute(sql)\n",
    "        \n",
    "#         sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "#         INTO TABLE locus_gene\n",
    "#         FIELDS TERMINATED BY '\\t'\n",
    "#         LINES TERMINATED BY '\\n'\n",
    "#         IGNORE 1 LINES;\"\"\".format(path = locus_gene_path)\n",
    "#         cursor.execute(sql)\n",
    "\n",
    "#         connection.commit()\n",
    "\n",
    "# finally:\n",
    "#     print(\"\\nDone\")\n",
    "#     connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">32 rows affected.</span>"
      ],
      "text/plain": [
       "32 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>germplasm</th>\n",
       "            <th>phenotype</th>\n",
       "            <th>pubmed</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AT1G01040</td>\n",
       "            <td>CS3828</td>\n",
       "            <td>Increased abundance of miRNA precursors.</td>\n",
       "            <td>17369351</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01060</td>\n",
       "            <td>lhy-101</td>\n",
       "            <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "            <td>16891401</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01140</td>\n",
       "            <td>SALK_058629</td>\n",
       "            <td>hypersensitive to low potassium media</td>\n",
       "            <td>17486125</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01220</td>\n",
       "            <td>SALK_012400C</td>\n",
       "            <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "            <td>18199744</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03720</td>\n",
       "            <td>SALK_042433</td>\n",
       "            <td>Multiple straight hairs</td>\n",
       "            <td>16367956</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03800</td>\n",
       "            <td>gek1-1</td>\n",
       "            <td>Ethanol hypersensitivity.</td>\n",
       "            <td>15215505</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G04240</td>\n",
       "            <td>xerico</td>\n",
       "            <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "            <td>17933900</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G05210</td>\n",
       "            <td>pot1-1</td>\n",
       "            <td>No visible phenotype.</td>\n",
       "            <td>17627276</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02130</td>\n",
       "            <td>rpk2-2</td>\n",
       "            <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "            <td>17419837</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02140</td>\n",
       "            <td>afp4-1</td>\n",
       "            <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "            <td>18484180</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>\n",
       "<span style=\"font-style:italic;text-align:center;\">Truncated to <a href=\"https://jupysql.ploomber.io/en/latest/api/configuration.html#displaylimit\">displaylimit</a> of 10.</span>"
      ],
      "text/plain": [
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "|   locus   |  germplasm   |                                                                                                                                                                      phenotype                                                                                                                                                                      |  pubmed  |\n",
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "| AT1G01040 |    CS3828    |                                                                                                                                                       Increased abundance of miRNA precursors.                                                                                                                                                      | 17369351 |\n",
       "| AT1G01060 |   lhy-101    |                                                                        The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.                                                                        | 16891401 |\n",
       "| AT1G01140 | SALK_058629  |                                                                                                                                                        hypersensitive to low potassium media                                                                                                                                                        | 17486125 |\n",
       "| AT1G01220 | SALK_012400C | fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected. | 18199744 |\n",
       "| AT2G03720 | SALK_042433  |                                                                                                                                                               Multiple straight hairs                                                                                                                                                               | 16367956 |\n",
       "| AT2G03800 |    gek1-1    |                                                                                                                                                              Ethanol hypersensitivity.                                                                                                                                                              | 15215505 |\n",
       "| AT2G04240 |    xerico    |                                                                                                                              Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.                                                                                                                             | 17933900 |\n",
       "| AT2G05210 |    pot1-1    |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 17627276 |\n",
       "| AT3G02130 |    rpk2-2    |                             The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.                             | 17419837 |\n",
       "| AT3G02140 |    afp4-1    |                                                                                                                                        Decreased germination on high concentrations of glucose and sorbitol.                                                                                                                                        | 18484180 |\n",
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "Truncated to displaylimit of 10."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">32 rows affected.</span>"
      ],
      "text/plain": [
       "32 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>gene</th>\n",
       "            <th>protein_lenght</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AT1G01040</td>\n",
       "            <td>DCL1</td>\n",
       "            <td>332</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01060</td>\n",
       "            <td>LHY</td>\n",
       "            <td>290</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01140</td>\n",
       "            <td>CIPK9</td>\n",
       "            <td>223</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01220</td>\n",
       "            <td>FKGP</td>\n",
       "            <td>190</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03720</td>\n",
       "            <td>MRH6</td>\n",
       "            <td>189</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03800</td>\n",
       "            <td>GEK1</td>\n",
       "            <td>196</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G04240</td>\n",
       "            <td>XERICO</td>\n",
       "            <td>256</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G05210</td>\n",
       "            <td>POT1A</td>\n",
       "            <td>221</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02130</td>\n",
       "            <td>RPK2</td>\n",
       "            <td>284</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02140</td>\n",
       "            <td>TMAC2</td>\n",
       "            <td>300</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>\n",
       "<span style=\"font-style:italic;text-align:center;\">Truncated to <a href=\"https://jupysql.ploomber.io/en/latest/api/configuration.html#displaylimit\">displaylimit</a> of 10.</span>"
      ],
      "text/plain": [
       "+-----------+--------+----------------+\n",
       "|   locus   |  gene  | protein_lenght |\n",
       "+-----------+--------+----------------+\n",
       "| AT1G01040 |  DCL1  |      332       |\n",
       "| AT1G01060 |  LHY   |      290       |\n",
       "| AT1G01140 | CIPK9  |      223       |\n",
       "| AT1G01220 |  FKGP  |      190       |\n",
       "| AT2G03720 |  MRH6  |      189       |\n",
       "| AT2G03800 |  GEK1  |      196       |\n",
       "| AT2G04240 | XERICO |      256       |\n",
       "| AT2G05210 | POT1A  |      221       |\n",
       "| AT3G02130 |  RPK2  |      284       |\n",
       "| AT3G02140 | TMAC2  |      300       |\n",
       "+-----------+--------+----------------+\n",
       "Truncated to displaylimit of 10."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def obtain_query_results(query: str, \n",
    "                        host: str = 'localhost', \n",
    "                        user: str = 'root',\n",
    "                        db: str = 'germplasm',\n",
    "                        charset: str = 'utf8mb4',\n",
    "                        local_infile: bool = True\n",
    "                        )-> list:\n",
    "    \"\"\"Returns the output from a mysql query in a single list. \n",
    "    The resulting list contains a dict for each orw of the output table,\n",
    "    Each dictionary row has a field_name:value key pair for each column of the table.\n",
    "    This function uses a pymsql DictCursor in the backend that a allows the conection with\n",
    "    the mysql server. Each parameter for the conection can be introduced.\n",
    "    WARNING: You should have configure a sqlalchemy conection with the server previously.\n",
    "    WARNING: For security reasons, it would ask for the server password each time.\n",
    "\n",
    "    :param: query: str - The mysql query.\n",
    "    :param: user: str - User of the mysql server for the connection configuration.\n",
    "    :param: db: str - Database to use.\n",
    "    :param: charset: str - Default utf8mb4. Character encoding configuration. \n",
    "    :local_infile: bool - Default True. Option to load local data files to the mysql database.\n",
    "    :return: list - List of dicts  containing all the values from the the output table.\n",
    "    \"\"\"    \n",
    "    \n",
    "    import pymysql.cursors\n",
    "    from getpass import getpass\n",
    "\n",
    "    connection = pymysql.connect(host = host,\n",
    "                                user = user,\n",
    "                                password = getpass(),\n",
    "                                db = db, \n",
    "                                charset = charset,\n",
    "                                local_infile = local_infile,\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "        \n",
    "            cursor.execute(query)\n",
    "\n",
    "            results = cursor.fetchall() # Fetch the results.\n",
    "            return results\n",
    "            \n",
    "    finally:\n",
    "        print(\"Query processed\")\n",
    "        connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_table_report(table: list, \n",
    "                          path: str = \"/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/report.tsv\"\n",
    "                          ):\n",
    "    \"\"\"Generates a report file from a table in the list of dict format generated by the obtain_query_results function.\n",
    "    It uses the same format as the cursor.fetchal() method from a pymysql DictCursor. \n",
    "    The file contains the table in an tsv format, including a header with the column names.\n",
    "\n",
    "    :param: table: list - A list of dicts containing the table.\n",
    "    :param: path: str - The path an file name for the report.\n",
    "    :return: file - .tsv file containing the formated content of the table.\"\"\"\n",
    "\n",
    "    file = open(path, 'a') # Uses append for non-destuctive writting.\n",
    "\n",
    "    # Print column names:\n",
    "    for column_name in table[0]: # Access the first dict in order to print the key names (which are the column names)\n",
    "        file.write(f'{column_name}\\t')\n",
    "\n",
    "    file.write(f'\\n') # New line for table content.\n",
    "\n",
    "    # Print table content:\n",
    "    for row in table: # Each element of the list is a row of the table.\n",
    "        for field in row.values(): # We loop on the values, as we already printed the column names (the keys).\n",
    "            file.write(f'{field}\\t')\n",
    "\n",
    "        file.write(f'\\n') # New line for each row.\n",
    "\n",
    "    file.write(f'\\n') # New line for each table.\n",
    "\n",
    "    print(f'Report generated at {path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at /home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/report.tsv\n"
     ]
    }
   ],
   "source": [
    "# TASK 1.\n",
    "\n",
    "query = \"\"\"SELECT * FROM germplasm \n",
    "            JOIN locus_gene USING (locus) \n",
    "            WHERE germplasm.locus = locus_gene.locus\"\"\"\n",
    "            \n",
    "generate_table_report(obtain_query_results(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at /home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/report.tsv\n"
     ]
    }
   ],
   "source": [
    "# TASK 2.\n",
    "\n",
    "query = \"\"\"SELECT * FROM germplasm \n",
    "            JOIN locus_gene USING (locus) \n",
    "            WHERE germplasm.locus = locus_gene.locus \n",
    "            AND locus_gene.gene = 'SKOR' OR locus_gene.gene = 'MAA3';\"\"\"\n",
    "            \n",
    "generate_table_report(obtain_query_results(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at /home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/report.tsv\n"
     ]
    }
   ],
   "source": [
    "# TASK 3.\n",
    "\n",
    "query = \"\"\"SELECT \n",
    "\n",
    "                (SELECT COUNT(*) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][1][G][0-9]{5}')\n",
    "                AS \"Entries for Chromosome 1\",\n",
    "\n",
    "                (SELECT COUNT(*) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][2][G][0-9]{5}')\n",
    "                AS \"Entries for Chromosome 2\",\n",
    "\n",
    "                (SELECT COUNT(*) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][3][G][0-9]{5}')\n",
    "                AS \"Entries for Chromosome 3\",\n",
    "\n",
    "                (SELECT COUNT(*) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][4][G][0-9]{5}')\n",
    "                AS \"Entries for Chromosome 4\",\n",
    "\n",
    "                (SELECT COUNT(*) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][5][G][0-9]{5}')\n",
    "                AS \"Entries for Chromosome 5\"\n",
    "                \n",
    "            ;\"\"\"\n",
    "            \n",
    "generate_table_report(obtain_query_results(query))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at test.tsv\n"
     ]
    }
   ],
   "source": [
    "# CHAT GPT ANSWER.\n",
    "\n",
    "query = \"\"\"SELECT \n",
    "    SUM(CASE WHEN locus REGEXP '[A][T][1][G][0-9]{5}' THEN 1 ELSE 0 END) AS \"Entries for Chromosome 1\",\n",
    "    SUM(CASE WHEN locus REGEXP '[A][T][2][G][0-9]{5}' THEN 1 ELSE 0 END) AS \"Entries for Chromosome 2\",\n",
    "    SUM(CASE WHEN locus REGEXP '[A][T][3][G][0-9]{5}' THEN 1 ELSE 0 END) AS \"Entries for Chromosome 3\",\n",
    "    SUM(CASE WHEN locus REGEXP '[A][T][4][G][0-9]{5}' THEN 1 ELSE 0 END) AS \"Entries for Chromosome 4\",\n",
    "    SUM(CASE WHEN locus REGEXP '[A][T][5][G][0-9]{5}' THEN 1 ELSE 0 END) AS \"Entries for Chromosome 5\"\n",
    "FROM germplasm\n",
    "JOIN locus_gene USING (locus);\"\"\"\n",
    "\n",
    "generate_table_report(table = obtain_query_results(query),\n",
    "                      path = \"test.tsv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at /home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/report.tsv\n"
     ]
    }
   ],
   "source": [
    "# TASK 4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx).\n",
    "\n",
    "query = \"\"\"SELECT \n",
    "\n",
    "                (SELECT AVG(protein_lenght) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][1][G][0-9]{5}')\n",
    "                AS \"Average protein length for Chromosome 1\",\n",
    "\n",
    "                (SELECT AVG(protein_lenght) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][2][G][0-9]{5}')\n",
    "                AS \"Average protein length for Chromosome 2\",\n",
    "\n",
    "                (SELECT AVG(protein_lenght) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][3][G][0-9]{5}')\n",
    "                AS \"Average protein length for Chromosome 3\",\n",
    "\n",
    "                (SELECT AVG(protein_lenght) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][4][G][0-9]{5}')\n",
    "                AS \"Average protein length for Chromosome 4\",\n",
    "\n",
    "                (SELECT AVG(protein_lenght) \n",
    "                    FROM germplasm\n",
    "                    JOIN locus_gene USING (locus)\n",
    "                    WHERE germplasm.locus = locus_gene.locus\n",
    "                    AND locus REGEXP '[A][T][5][G][0-9]{5}')\n",
    "                AS \"Average protein length for Chromosome 5\"\n",
    "                \n",
    "            ;\"\"\"\n",
    "            \n",
    "generate_table_report(obtain_query_results(query))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "Report generated at test.tsv\n"
     ]
    }
   ],
   "source": [
    "# CHAT GPT ANSWER.\n",
    "\n",
    "query = \"\"\"SELECT \n",
    "    AVG(CASE WHEN locus REGEXP '[A][T][1][G][0-9]{5}' THEN protein_lenght ELSE NULL END) AS \"Average protein length for Chromosome 1\",\n",
    "    AVG(CASE WHEN locus REGEXP '[A][T][2][G][0-9]{5}' THEN protein_lenght ELSE NULL END) AS \"Average protein length for Chromosome 2\",\n",
    "    AVG(CASE WHEN locus REGEXP '[A][T][3][G][0-9]{5}' THEN protein_lenght ELSE NULL END) AS \"Average protein length for Chromosome 3\",\n",
    "    AVG(CASE WHEN locus REGEXP '[A][T][4][G][0-9]{5}' THEN protein_lenght ELSE NULL END) AS \"Average protein length for Chromosome 4\",\n",
    "    AVG(CASE WHEN locus REGEXP '[A][T][5][G][0-9]{5}' THEN protein_lenght ELSE NULL END) AS \"Average protein length for Chromosome 5\"\n",
    "FROM germplasm\n",
    "JOIN locus_gene USING (locus);\"\"\"\n",
    "\n",
    "generate_table_report(table = obtain_query_results(query),\n",
    "                      path = \"test.tsv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "course",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
