{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Start your mySQL Server from a terminal \n",
    "(if it isn't already running)\n",
    "\n",
    "<code>sudo docker start course-mysql</code>\n",
    "<pre>\n",
    "\n",
    "</pre>\n",
    "Don't forget that, if you use sqlMagic, you need to execute the connection lines in your Notebook!\n",
    "\n",
    "<pre>\n",
    "%load_ext sql\n",
    "%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "</pre>\n",
    "\n",
    "## Create a new Python3 Jupyter Notebook in your Exam Answers folder\n",
    "\n",
    "Commit and push this Notebook to GitHub when you are finished.\n",
    "\n",
    "You must submit your answers to GitHub by 1800h Sept 14."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bash:\n",
    "\n",
    "# docker start course-mysql\n",
    "# mysql --local_infile=true -h localhost -P 3306 --protocol=TCP -u root -p "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mThere's a new jupysql version available (0.10.3), you're running 0.10.1. To upgrade: pip install jupysql --upgrade\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "++\n",
       "||\n",
       "++\n",
       "++"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from sqlalchemy import URL, create_engine, MetaData\n",
    "from getpass import getpass\n",
    "\n",
    "url_object = URL.create(\n",
    "    \"mysql+pymysql\",\n",
    "    username=\"root\",\n",
    "    password= getpass(),  # plain (unescaped) text\n",
    "    host=\"localhost\", # Averiguar en la línea de comandos con \"status;\". Ni idea de porqué no hace ni puto caso al comando de antes.\n",
    "    database=\"mysql\"\n",
    ")\n",
    "\n",
    "engine = create_engine(url_object)\n",
    "\n",
    "metadata = MetaData() # Crea una variable para almacenar los metadatos de las tablas.\n",
    "\n",
    "# Inicia el interprete de pymysql.\n",
    "%load_ext sql \n",
    "\n",
    "# Realiza la conexión.\n",
    "%sql engine \n",
    "\n",
    "# Permite cargar archivos locales.\n",
    "%sql SET GLOBAL local_infile=true;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Data Files\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Problem 1:  Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx).  (hint: This will help you decide how to load the data into the database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "def get_first_column(\n",
    "    file_path: str, \n",
    "    header: bool = False, \n",
    "    delimiter: str = '\\t', \n",
    "    quotechar: str = '\"'\n",
    "    ) -> list: \n",
    "    \"\"\"Return the first column of a data file in a single list. \n",
    "    Each value is imported as a string.\n",
    "\n",
    "    :param: file_name: str - Path to the file containing the data.\n",
    "    :param: header: bool - Default false. True if the file contains a \n",
    "    header and should be ignored. \n",
    "    :param: delimiter: str - Default tab (.tsv file). Delimiter for \n",
    "    the fields it the data file.\n",
    "    :param: quotechar: str - Default '\"'. Quoting for the imported data.\n",
    "    :return: list - List containing all the values from the first \n",
    "    column of the file.\n",
    "    \"\"\"    \n",
    "\n",
    "    import csv\n",
    "\n",
    "    first_column = []\n",
    "    rest = []\n",
    "\n",
    "    with open(file_path) as file:\n",
    "\n",
    "        if header: next(file) # Es un iterable así que nos saltamos la cabecera.\n",
    "        \n",
    "        file_reader = csv.reader(file, delimiter = delimiter, quotechar = quotechar)\n",
    "        \n",
    "        for first_column_values, *rest in file_reader: \n",
    "            # Itera por filas y automáticamente guarda el valor de cada \n",
    "            # campo en cada una de las variables.\n",
    "            first_column.append(first_column_values)\n",
    "            rest.clear() # Rest no lo itero así que solo guardará el último valor.\n",
    "\n",
    "    return first_column\n",
    "\n",
    "\n",
    "locus_Germplasm = get_first_column(\"Germplasm.tsv\", True)\n",
    "locus_LocusGene = get_first_column(\"LocusGene.tsv\", True)\n",
    "\n",
    "print(locus_Germplasm == locus_LocusGene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">1 rows affected.</span>"
      ],
      "text/plain": [
       "1 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">5 rows affected.</span>"
      ],
      "text/plain": [
       "5 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>Tables_in_germplasm</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+---------------------+\n",
       "| Tables_in_germplasm |\n",
       "+---------------------+\n",
       "+---------------------+"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql drop database germplasm;\n",
    "%sql create database germplasm;\n",
    "%sql show databases;\n",
    "%sql use germplasm;\n",
    "%sql show tables;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password=getpass(),\n",
    "                             db='germplasm', \n",
    "                             charset='utf8mb4',\n",
    "                             local_infile=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        \n",
    "        sql = \"\"\"create table germplasm(\n",
    "            locus VARCHAR(20) NOT NULL PRIMARY KEY,\n",
    "            germplasm VARCHAR(100) NOT NULL,\n",
    "            phenotype VARCHAR(1000) NOT NULL,\n",
    "            pubmed INTEGER NOT NULL);\"\"\"\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        sql = \"\"\"create table locus_gene(\n",
    "            locus VARCHAR(20) NOT NULL PRIMARY KEY,\n",
    "            gene VARCHAR(20) NOT NULL,\n",
    "            protein_lenght INTEGER NOT NULL)\"\"\"\n",
    "        cursor.execute(sql)\n",
    "\n",
    "    connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\\nDone\")\n",
    "    connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">2 rows affected.</span>"
      ],
      "text/plain": [
       "2 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>Tables_in_germplasm</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>germplasm</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>locus_gene</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+---------------------+\n",
       "| Tables_in_germplasm |\n",
       "+---------------------+\n",
       "|      germplasm      |\n",
       "|      locus_gene     |\n",
       "+---------------------+"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show tables;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>germplasm</th>\n",
       "            <th>phenotype</th>\n",
       "            <th>pubmed</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+-------+-----------+-----------+--------+\n",
       "| locus | germplasm | phenotype | pubmed |\n",
       "+-------+-----------+-----------+--------+\n",
       "+-------+-----------+-----------+--------+"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>gene</th>\n",
       "            <th>protein_lenght</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "+-------+------+----------------+\n",
       "| locus | gene | protein_lenght |\n",
       "+-------+------+----------------+\n",
       "+-------+------+----------------+"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done\n"
     ]
    },
    {
     "ename": "OperationalError",
     "evalue": "(1017, \"Can't find file 'b'/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv''\")",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:1415\u001b[0m, in \u001b[0;36mLoadLocalFile.send_data\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1414\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1415\u001b[0m     \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfilename, \u001b[39m\"\u001b[39;49m\u001b[39mrb\u001b[39;49m\u001b[39m\"\u001b[39;49m) \u001b[39mas\u001b[39;00m open_file:\n\u001b[1;32m   1416\u001b[0m         packet_size \u001b[39m=\u001b[39m \u001b[39mmin\u001b[39m(\n\u001b[1;32m   1417\u001b[0m             conn\u001b[39m.\u001b[39mmax_allowed_packet, \u001b[39m16\u001b[39m \u001b[39m*\u001b[39m \u001b[39m1024\u001b[39m\n\u001b[1;32m   1418\u001b[0m         )  \u001b[39m# 16KB is efficient enough\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: b'/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mOperationalError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[1;32m/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb Cell 14\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=15'>16</a>\u001b[0m locus_gene_path \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/LocusGene.tsv\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=17'>18</a>\u001b[0m sql \u001b[39m=\u001b[39m \u001b[39m\"\"\"\u001b[39m\u001b[39mLOAD DATA LOCAL INFILE \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{path}\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=18'>19</a>\u001b[0m \u001b[39mINTO TABLE germplasm\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=19'>20</a>\u001b[0m \u001b[39mFIELDS TERMINATED BY \u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=20'>21</a>\u001b[0m \u001b[39mLINES TERMINATED BY \u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=21'>22</a>\u001b[0m \u001b[39mIGNORE 1 LINES;\u001b[39m\u001b[39m\"\"\"\u001b[39m\u001b[39m.\u001b[39mformat(path \u001b[39m=\u001b[39m germplasm_path)\n\u001b[0;32m---> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=22'>23</a>\u001b[0m cursor\u001b[39m.\u001b[39;49mexecute(sql)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=24'>25</a>\u001b[0m sql \u001b[39m=\u001b[39m \u001b[39m\"\"\"\u001b[39m\u001b[39mLOAD DATA LOCAL INFILE \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{path}\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=25'>26</a>\u001b[0m \u001b[39mINTO TABLE locus_gene\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=26'>27</a>\u001b[0m \u001b[39mFIELDS TERMINATED BY \u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\t\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=27'>28</a>\u001b[0m \u001b[39mLINES TERMINATED BY \u001b[39m\u001b[39m'\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m'\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=28'>29</a>\u001b[0m \u001b[39mIGNORE 1 LINES;\u001b[39m\u001b[39m\"\"\"\u001b[39m\u001b[39m.\u001b[39mformat(path \u001b[39m=\u001b[39m locus_gene_path)\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/jm/Course/Accelerated_Intro_WilkinsonExams/Exam_2_Answers.ipynb#X16sdnNjb2RlLXJlbW90ZQ%3D%3D?line=29'>30</a>\u001b[0m cursor\u001b[39m.\u001b[39mexecute(sql)\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/cursors.py:153\u001b[0m, in \u001b[0;36mCursor.execute\u001b[0;34m(self, query, args)\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[39mpass\u001b[39;00m\n\u001b[1;32m    151\u001b[0m query \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmogrify(query, args)\n\u001b[0;32m--> 153\u001b[0m result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_query(query)\n\u001b[1;32m    154\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_executed \u001b[39m=\u001b[39m query\n\u001b[1;32m    155\u001b[0m \u001b[39mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/cursors.py:322\u001b[0m, in \u001b[0;36mCursor._query\u001b[0;34m(self, q)\u001b[0m\n\u001b[1;32m    320\u001b[0m conn \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_get_db()\n\u001b[1;32m    321\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_clear_result()\n\u001b[0;32m--> 322\u001b[0m conn\u001b[39m.\u001b[39;49mquery(q)\n\u001b[1;32m    323\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_do_get_result()\n\u001b[1;32m    324\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mrowcount\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:558\u001b[0m, in \u001b[0;36mConnection.query\u001b[0;34m(self, sql, unbuffered)\u001b[0m\n\u001b[1;32m    556\u001b[0m     sql \u001b[39m=\u001b[39m sql\u001b[39m.\u001b[39mencode(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mencoding, \u001b[39m\"\u001b[39m\u001b[39msurrogateescape\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    557\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_execute_command(COMMAND\u001b[39m.\u001b[39mCOM_QUERY, sql)\n\u001b[0;32m--> 558\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_affected_rows \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_query_result(unbuffered\u001b[39m=\u001b[39;49munbuffered)\n\u001b[1;32m    559\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_affected_rows\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:822\u001b[0m, in \u001b[0;36mConnection._read_query_result\u001b[0;34m(self, unbuffered)\u001b[0m\n\u001b[1;32m    820\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    821\u001b[0m     result \u001b[39m=\u001b[39m MySQLResult(\u001b[39mself\u001b[39m)\n\u001b[0;32m--> 822\u001b[0m     result\u001b[39m.\u001b[39;49mread()\n\u001b[1;32m    823\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_result \u001b[39m=\u001b[39m result\n\u001b[1;32m    824\u001b[0m \u001b[39mif\u001b[39;00m result\u001b[39m.\u001b[39mserver_status \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:1205\u001b[0m, in \u001b[0;36mMySQLResult.read\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1203\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_read_ok_packet(first_packet)\n\u001b[1;32m   1204\u001b[0m \u001b[39melif\u001b[39;00m first_packet\u001b[39m.\u001b[39mis_load_local_packet():\n\u001b[0;32m-> 1205\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_read_load_local_packet(first_packet)\n\u001b[1;32m   1206\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1207\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_read_result_packet(first_packet)\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:1253\u001b[0m, in \u001b[0;36mMySQLResult._read_load_local_packet\u001b[0;34m(self, first_packet)\u001b[0m\n\u001b[1;32m   1251\u001b[0m sender \u001b[39m=\u001b[39m LoadLocalFile(load_packet\u001b[39m.\u001b[39mfilename, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mconnection)\n\u001b[1;32m   1252\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1253\u001b[0m     sender\u001b[39m.\u001b[39;49msend_data()\n\u001b[1;32m   1254\u001b[0m \u001b[39mexcept\u001b[39;00m:\n\u001b[1;32m   1255\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mconnection\u001b[39m.\u001b[39m_read_packet()  \u001b[39m# skip ok packet\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/course/lib/python3.11/site-packages/pymysql/connections.py:1425\u001b[0m, in \u001b[0;36mLoadLocalFile.send_data\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1423\u001b[0m             conn\u001b[39m.\u001b[39mwrite_packet(chunk)\n\u001b[1;32m   1424\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mOSError\u001b[39;00m:\n\u001b[0;32m-> 1425\u001b[0m     \u001b[39mraise\u001b[39;00m err\u001b[39m.\u001b[39mOperationalError(\n\u001b[1;32m   1426\u001b[0m         ER\u001b[39m.\u001b[39mFILE_NOT_FOUND,\n\u001b[1;32m   1427\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCan\u001b[39m\u001b[39m'\u001b[39m\u001b[39mt find file \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m   1428\u001b[0m     )\n\u001b[1;32m   1429\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m   1430\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m conn\u001b[39m.\u001b[39m_closed:\n\u001b[1;32m   1431\u001b[0m         \u001b[39m# send the empty packet to signify we are done sending data\u001b[39;00m\n",
      "\u001b[0;31mOperationalError\u001b[0m: (1017, \"Can't find file 'b'/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv''\")"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password=getpass(),\n",
    "                             db='germplasm', \n",
    "                             charset='utf8mb4',\n",
    "                             local_infile=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        germplasm_path = \"/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv\"\n",
    "        locus_gene_path = \"/home/usuario/Documents/Introduction_course/Accelerated_Intro_WilkinsonExams/LocusGene.tsv\"\n",
    "\n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE germplasm\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = germplasm_path)\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE locus_gene\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = locus_gene_path)\n",
    "        cursor.execute(sql)\n",
    "\n",
    "        connection.commit()\n",
    "\n",
    "finally:\n",
    "    print(\"\\nDone\")\n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "#  PARA EL PC PERSONAL\n",
    "\n",
    "import pymysql.cursors\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password=getpass(),\n",
    "                             db='germplasm', \n",
    "                             charset='utf8mb4',\n",
    "                             local_infile=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "\n",
    "        germplasm_path = \"/home/jm/Course/Accelerated_Intro_WilkinsonExams/Germplasm.tsv\"\n",
    "        locus_gene_path = \"/home/jm/Course/Accelerated_Intro_WilkinsonExams/LocusGene.tsv\"\n",
    "\n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE germplasm\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = germplasm_path)\n",
    "        cursor.execute(sql)\n",
    "        \n",
    "        sql = \"\"\"LOAD DATA LOCAL INFILE '{path}'\n",
    "        INTO TABLE locus_gene\n",
    "        FIELDS TERMINATED BY '\\t'\n",
    "        LINES TERMINATED BY '\\n'\n",
    "        IGNORE 1 LINES;\"\"\".format(path = locus_gene_path)\n",
    "        cursor.execute(sql)\n",
    "\n",
    "        connection.commit()\n",
    "\n",
    "finally:\n",
    "    print(\"\\nDone\")\n",
    "    connection.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">32 rows affected.</span>"
      ],
      "text/plain": [
       "32 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>germplasm</th>\n",
       "            <th>phenotype</th>\n",
       "            <th>pubmed</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AT1G01040</td>\n",
       "            <td>CS3828</td>\n",
       "            <td>Increased abundance of miRNA precursors.</td>\n",
       "            <td>17369351</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01060</td>\n",
       "            <td>lhy-101</td>\n",
       "            <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "            <td>16891401</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01140</td>\n",
       "            <td>SALK_058629</td>\n",
       "            <td>hypersensitive to low potassium media</td>\n",
       "            <td>17486125</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01220</td>\n",
       "            <td>SALK_012400C</td>\n",
       "            <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "            <td>18199744</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03720</td>\n",
       "            <td>SALK_042433</td>\n",
       "            <td>Multiple straight hairs</td>\n",
       "            <td>16367956</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03800</td>\n",
       "            <td>gek1-1</td>\n",
       "            <td>Ethanol hypersensitivity.</td>\n",
       "            <td>15215505</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G04240</td>\n",
       "            <td>xerico</td>\n",
       "            <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "            <td>17933900</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G05210</td>\n",
       "            <td>pot1-1</td>\n",
       "            <td>No visible phenotype.</td>\n",
       "            <td>17627276</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02130</td>\n",
       "            <td>rpk2-2</td>\n",
       "            <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "            <td>17419837</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02140</td>\n",
       "            <td>afp4-1</td>\n",
       "            <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "            <td>18484180</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>\n",
       "<span style=\"font-style:italic;text-align:center;\">Truncated to <a href=\"https://jupysql.ploomber.io/en/latest/api/configuration.html#displaylimit\">displaylimit</a> of 10.</span>"
      ],
      "text/plain": [
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "|   locus   |  germplasm   |                                                                                                                                                                      phenotype                                                                                                                                                                      |  pubmed  |\n",
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "| AT1G01040 |    CS3828    |                                                                                                                                                       Increased abundance of miRNA precursors.                                                                                                                                                      | 17369351 |\n",
       "| AT1G01060 |   lhy-101    |                                                                        The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.                                                                        | 16891401 |\n",
       "| AT1G01140 | SALK_058629  |                                                                                                                                                        hypersensitive to low potassium media                                                                                                                                                        | 17486125 |\n",
       "| AT1G01220 | SALK_012400C | fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected. | 18199744 |\n",
       "| AT2G03720 | SALK_042433  |                                                                                                                                                               Multiple straight hairs                                                                                                                                                               | 16367956 |\n",
       "| AT2G03800 |    gek1-1    |                                                                                                                                                              Ethanol hypersensitivity.                                                                                                                                                              | 15215505 |\n",
       "| AT2G04240 |    xerico    |                                                                                                                              Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.                                                                                                                             | 17933900 |\n",
       "| AT2G05210 |    pot1-1    |                                                                                                                                                                No visible phenotype.                                                                                                                                                                | 17627276 |\n",
       "| AT3G02130 |    rpk2-2    |                             The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.                             | 17419837 |\n",
       "| AT3G02140 |    afp4-1    |                                                                                                                                        Decreased germination on high concentrations of glucose and sorbitol.                                                                                                                                        | 18484180 |\n",
       "+-----------+--------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------+\n",
       "Truncated to displaylimit of 10."
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM germplasm;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"None\">Running query in &#x27;mysql+pymysql://root:***@localhost/mysql&#x27;</span>"
      ],
      "text/plain": [
       "Running query in 'mysql+pymysql://root:***@localhost/mysql'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color: green\">32 rows affected.</span>"
      ],
      "text/plain": [
       "32 rows affected."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>locus</th>\n",
       "            <th>gene</th>\n",
       "            <th>protein_lenght</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AT1G01040</td>\n",
       "            <td>DCL1</td>\n",
       "            <td>332</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01060</td>\n",
       "            <td>LHY</td>\n",
       "            <td>290</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01140</td>\n",
       "            <td>CIPK9</td>\n",
       "            <td>223</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT1G01220</td>\n",
       "            <td>FKGP</td>\n",
       "            <td>190</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03720</td>\n",
       "            <td>MRH6</td>\n",
       "            <td>189</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G03800</td>\n",
       "            <td>GEK1</td>\n",
       "            <td>196</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G04240</td>\n",
       "            <td>XERICO</td>\n",
       "            <td>256</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT2G05210</td>\n",
       "            <td>POT1A</td>\n",
       "            <td>221</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02130</td>\n",
       "            <td>RPK2</td>\n",
       "            <td>284</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AT3G02140</td>\n",
       "            <td>TMAC2</td>\n",
       "            <td>300</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>\n",
       "<span style=\"font-style:italic;text-align:center;\">Truncated to <a href=\"https://jupysql.ploomber.io/en/latest/api/configuration.html#displaylimit\">displaylimit</a> of 10.</span>"
      ],
      "text/plain": [
       "+-----------+--------+----------------+\n",
       "|   locus   |  gene  | protein_lenght |\n",
       "+-----------+--------+----------------+\n",
       "| AT1G01040 |  DCL1  |      332       |\n",
       "| AT1G01060 |  LHY   |      290       |\n",
       "| AT1G01140 | CIPK9  |      223       |\n",
       "| AT1G01220 |  FKGP  |      190       |\n",
       "| AT2G03720 |  MRH6  |      189       |\n",
       "| AT2G03800 |  GEK1  |      196       |\n",
       "| AT2G04240 | XERICO |      256       |\n",
       "| AT2G05210 | POT1A  |      221       |\n",
       "| AT3G02130 |  RPK2  |      284       |\n",
       "| AT3G02140 | TMAC2  |      300       |\n",
       "+-----------+--------+----------------+\n",
       "Truncated to displaylimit of 10."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT * FROM locus_gene;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query processed\n",
      "[{'locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'gene': 'DCL1', 'protein_lenght': 332}, {'locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'gene': 'LHY', 'protein_lenght': 290}, {'locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'gene': 'CIPK9', 'protein_lenght': 223}, {'locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'gene': 'FKGP', 'protein_lenght': 190}, {'locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'gene': 'MRH6', 'protein_lenght': 189}, {'locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'gene': 'GEK1', 'protein_lenght': 196}, {'locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'gene': 'XERICO', 'protein_lenght': 256}, {'locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'gene': 'POT1A', 'protein_lenght': 221}, {'locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'gene': 'RPK2', 'protein_lenght': 284}, {'locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'gene': 'TMAC2', 'protein_lenght': 300}, {'locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'gene': 'RGP1', 'protein_lenght': 301}, {'locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'gene': 'BIG', 'protein_lenght': 279}, {'locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'gene': 'SEP2', 'protein_lenght': 175}, {'locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'gene': 'NBS1', 'protein_lenght': 190}, {'locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'gene': 'SKOR', 'protein_lenght': 234}, {'locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'gene': 'VTC4', 'protein_lenght': 311}, {'locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'gene': 'HDG8', 'protein_lenght': 194}, {'locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'gene': 'SUV3', 'protein_lenght': 312}, {'locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'gene': 'BAM5', 'protein_lenght': 313}, {'locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'gene': 'DXS', 'protein_lenght': 219}, {'locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'gene': 'MAA3', 'protein_lenght': 294}, {'locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'gene': 'HSBP', 'protein_lenght': 254}, {'locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'gene': 'ESD4', 'protein_lenght': 265}, {'locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'gene': 'ADA2B', 'protein_lenght': 279}, {'locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'gene': 'INT4', 'protein_lenght': 284}, {'locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'gene': 'PAS2', 'protein_lenght': 301}, {'locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'gene': 'AIL6', 'protein_lenght': 310}, {'locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'gene': 'SPS2', 'protein_lenght': 232}, {'locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'gene': 'HY5', 'protein_lenght': 221}, {'locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'gene': 'MYB3R-4', 'protein_lenght': 336}, {'locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'gene': 'PYD2', 'protein_lenght': 310}, {'locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'gene': 'CRN', 'protein_lenght': 189}]\n"
     ]
    }
   ],
   "source": [
    "def obtain_query_results(query: str, \n",
    "                        host: str = 'localhost', \n",
    "                        user: str = 'root',\n",
    "                        password: str = getpass(),\n",
    "                        db: str = 'germplasm',\n",
    "                        charset: str = 'utf8mb4',\n",
    "                        local_infile: bool = True\n",
    "                        )-> list:\n",
    "    \n",
    "    import pymysql.cursors\n",
    "\n",
    "    connection = pymysql.connect(host = host,\n",
    "                                user = user,\n",
    "                                password = password,\n",
    "                                db = db, \n",
    "                                charset = charset,\n",
    "                                local_infile = local_infile,\n",
    "                                cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "        \n",
    "            cursor.execute(query)\n",
    "\n",
    "            results = cursor.fetchall() # Fetch the results.\n",
    "            return results\n",
    "            \n",
    "    finally:\n",
    "        print(\"Query processed\")\n",
    "        connection.close()\n",
    "\n",
    "query = \"\"\"SELECT * FROM germplasm \n",
    "            JOIN locus_gene USING (locus) \n",
    "            WHERE germplasm.locus = locus_gene.locus\"\"\"\n",
    "            \n",
    "result = obtain_query_results(query)\n",
    "print(result)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "locus\tgermplasm\tphenotype\tpubmed\tgene\tprotein_lenght\t\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\t\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\t\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\t\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\t\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\t\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\t\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\t\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\t\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\t\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\t\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\t\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\t\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\t\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\t\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\t\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\t\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\t\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\t\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\t\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\t\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\t\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\t\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\t\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\t\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\t\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\t\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\t\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\t\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\t\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\t\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\t\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\t\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def generate_table_report(table: list):\n",
    "\n",
    "    # Print column names:\n",
    "    for column_name in table[0]: # Access the first dict in order to print the key names (which are the column names)\n",
    "        print(column_name, end = '\\t')\n",
    "\n",
    "    print(\"\") # New line for table content.\n",
    "\n",
    "    # Print table content:\n",
    "    for row in table: # Each element of the list is a row of the table.\n",
    "        for field in row.values(): # We loop on the values, as we already printed the column names (the keys).\n",
    "            print(field, end = '\\t')\n",
    "\n",
    "        print(\"\") # New line for each row.\n",
    "\n",
    "    print(\"\") # New line for each table.\n",
    "\n",
    "generate_table_report(result)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "course",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
